### ISSUE TITLE
Rough Roadmap

### ISSUE THREAD

**Issue by CaptainProton42** [CaptainProton42/GodotHoloPlayExtension#3]
This is a rough roadmap for porting the plugin from 3.x to an extension for 4.0. Will probably change a lot and input is appreciated!

- [ ] Implement `HoloPlayManager` singleton class
   - This class should  manage all connected HoloPlay devices, e.g. keep a list of connected devices so we can supply all `HoloPlayVolumes` with a dropdown list. Since it's a singleton it can also manage static content like the compiled lightfield shader. We could also give the user access to state functions such as `hpc_InitializeApp` but this might be more inconvenient than useful.
   - [ ] Poll for devices using ~`hpc_RefreshState` and~ `hpc_GetNumDevices` etc. at sensible intervals/times.
      - [ ] Maybe when a `HoloPlayVolume` node is instantiated/selected in the editor or when the device list dropdown is accessed in the editor? Any ideas?
   - [x]  Use `RenderingDevice` to compile the lightfield shader (shaders require some run-time tweaking to make them compile in Vulkan)
   - [x] Retrieve a list of all connected devices (might work with `hpc_GetNumDevices` and iteration using `hpc_GetDevice*`
      - `Dictionary` support is not complete in the current C++ bindings so this could still be improved to contain more info.
   - [x] Initialise and close the app on the HoloPlay service
      - `initialize_app` compiles the shaders and initialises the app with the HoloPlay Service. `hpc_close` is called automatically.
   - [ ] Implement state functions
      - [x] `get_holoplay_core_version`
      - [ ] ...
   - [ ] ...
- [ ] Implement `HoloPlayVolume` class
   - Basically the same as the `HoloPlayVolume` class from the `3.x` plugin but does not store static data such as the compiled shader. 
   - [x] Inherits `Node3D`
   - [x] Retrieve a list of available devices from `HoloPlayManager`
   - [x] Use `DisplayServer` for windowing logic (no platform-specific code for context sharing and GLFW needed anymore.)
   - [ ] Use `RenderingDevice` or whichever class is used for this in Godot 4.0 to render the quilt using the compiled shader from `HoloPlayManager`
   - [ ] ...
- [ ] Reimplement editor gizmo.
   - [ ] Should be a relatively straight port from the 3.x version.
- [ ] Reimplement `HoloPlayRecorder` node
   - [ ] Can also be ported from 3.x althought this could probably use some improvements as well.

**Comment by fire** [comment]
Happy New Year and to HoloPlay4

**Comment by CaptainProton42** [comment]
Thanks! I should make finding more time for this again my new year's resolution! 😅

**Comment by fire** [comment]
@CaptainProton42 How can I help?

**Comment by trigger7** [comment]
Hello,

First off, thanks a lot for your work so far !

Now that Godot 4 has had several beta releases, are you planning to pick this project up again, or should someone else take over (either as a new project, or as contribution to this one) ?

No pressure either way :)

**Comment by CaptainProton42** [comment]
@trigger7 As it currently stands I have pretty much abandoned this project due to time constraints and just general lack of motivation. I don't really know if I'll pick it up now or in the future. 

But it's fully MIT licensed, so feel free to pick this up on your own fork or make a new project! (And let me know if you do!)

**Comment by trigger7** [comment]
Thanks for your response !
Fair enough, I totally understand.

I may take a shot at it then, but I'm very unfamiliar with this kind of stuff (and my cpp's not great...). And it's going to be low priority for me too, I won't have much time to dedicate to this.

If I do anything, it'll be in a fork of this repo. And in the off-chance I end up with anything useful, I'll make a pull request back here.

**Comment by shiena** [comment]
The Looking Glass Bridge SDK has been released to replace the Looking Glass Core SDK. Are there any plans to support this SDK?
https://docs.lookingglassfactory.com/core/looking-glass-bridge-sdk
